### PR TITLE
NotificationServer: Expand the notification when hovered + friends

### PR DESCRIPTION
**Pull request by sin-ack**
**LibGUI: Let GUI applications create an inspectable EventLoop**

GUI applications automatically have `EventLoop`s created for them via
`GUI::Application`; however before this commit it was not possible to make
the appliaction event loop inspectable. This exposes a third optional
argument to `GUI::Application` which allows one to make the application
event loop inspectable.

**LibGUI: Add virtual handlers for `WindowEntered` and `WindowLeft` events**

These can be useful if an application wants to react to the cursor
entering the window at any point, rather than just on a widget.

**LibGUI: Only dispatch `Leave` if the now-hovered widget isn't a child**

Without this change, when the mouse starts hovering over a child widget,
the parent would receive a Leave event despite the parent widget not
losing mouse hover.

**LibGUI: Ignore the `Enter` event by default**

If this is not ignored by default, then one can move the mouse fast
enough to skip hovering over a parent element and hover directly over
the child, causing the parent to never receive an `Enter` event.

**Userland: Move text wrapping/elision into the new TextLayout :^)**

This class now contains all the fun bits about laying out text in a
rect. It will handle line wrapping at a certain width, cutting off lines
that don't fit the given rect, and handling text elision.
`Painter::draw_text` now internally uses this.

Future work here would be not laying out text twice (once actually
preparing the lines to be rendered and once to get the bounding box),
and possibly adding left elision if necessary.

Additionally, this commit makes the `Utf32View` versions of
`Painter::draw_text` convert to `Utf8View` internally. The intention is to
completely remove those versions, but they're kept at the moment to keep
the scope of this PR small.
    
**NotificationServer: Expand the notification when hovered**

Now, instead of showing a tooltip, the entire notification will be
shown when the user hovers over a notification. In the future, limiting
the amount of lines shown in the notification might be a good idea.

### CODE REVIEW

File: Userland/Libraries/LibGUI/Label.cpp [hunk 1]
```diff
@@ -131,58 +107,11 @@ void Label::size_to_fit()
...
-        line_width += word_width;
-    }
-
-    auto last_line = builder.to_string();
-    if (!last_line.is_empty())
-        lines.append(last_line);
-
-    m_lines = lines;
+    // FIXME: The 4 is taken from Gfx::Painter and should be available as
+    // a constant instead.
```
Comment (alimpfard) [inline Userland/Libraries/LibGUI/Label.cpp:113]: Nit: please indent the second and later lines in these kinds of comments until after the `:`
```suggestion
    // FIXME: The 4 is taken from Gfx::Painter and should be available as
    //        a constant instead.
```

This makes (at least) CLion highlight the entire block as a FIXME and not just the first line.

File: Userland/Libraries/LibGfx/TextLayout.cpp [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        switch (*it) {
+        case '\n':
+        case '\r':
+        case '\t':
+        case ' ': {
+            if (start_byte_offset.has_value())
+                words.append(m_text.substring_view(start_byte_offset.value(), current_byte_offset - start_byte_offset.value()));
+            start_byte_offset.clear();
+
+            // FIXME: is there anyway we can get Mac newlines (\r) in text?
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/TextLayout.cpp:53]: I don't understand what this FIXME defers.
"FIXME: Add support for mac-style newlines (\r)"? not sure why that would be necessary or ever happen.

Comment (sin-ack) [reply Userland/Libraries/LibGfx/TextLayout.cpp:53]: Very good question. I don't remember writing this.

**Review by alimpfard — APPROVED** [review]
nice branch.
LGTM

**Pull request merged**